src/sqlfluff/core/rules/surrogates/segments.py

```diff
@@ -0,0 +1,130 @@
+"""Surrogate class for working with Segment collections."""
+from typing import Any, Callable, List, Optional, Sequence, Type, TypeVar
+
+from sqlfluff.core.parser import BaseSegment
+from sqlfluff.core.templaters.base import TemplatedFile
+from sqlfluff.core.rules.surrogates.raw_file_slice import RawFileSlices
+
+Predicate = TypeVar("Predicate", str, Type, Callable[[BaseSegment], bool])
+
+
+class Segments(tuple):
+    """Encapsulates a sequence of one or more BaseSegments.
+
+    The segments may or may not be contiguous in a parse tree.
+    Provides useful operations on a sequence of segments to simplify rule creation.
+    """
+
+    def __new__(cls, templated_file, *segments):
+        """Override new operator."""
+        return super(Segments, cls).__new__(cls, segments)
+
+    def __init__(self, templated_file: Optional[TemplatedFile], *_: BaseSegment):
+        self.templated_file = templated_file
+
+    def all(self, *predicates: Predicate) -> bool:
+        """Do all the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return all(cp(s) for s in self)
+
+    def any(self, *predicates: Predicate) -> bool:  # pragma: no cover
+        """Do any of the segments match?"""
+        cp = _CompositePredicate(*predicates)
+        return any(cp(s) for s in self)
+
+    @property
+    def raw_slices(self) -> RawFileSlices:
+        """Raw slices of the segments."""
+        if not self.templated_file:  # pragma: no cover
+            raise ValueError(
+                'Segments.raw_slices: "templated_file" property is required.'
+            )
+        raw_slices = set()
+        for s in self:
+            source_slice = s.pos_marker.source_slice
+            raw_slices.update(
+                self.templated_file.raw_slices_spanning_source_slice(source_slice)
+            )
+        return RawFileSlices(self.templated_file, *raw_slices)
+
+    def children(self, *predicates: Predicate) -> "Segments":
+        """Returns an object with children of the segments in this object."""
+        child_segments: List[BaseSegment] = []
+        cp = _CompositePredicate(*predicates)
+        for s in self:
+            for child in s.segments:
+                if cp(child):
+                    child_segments.append(child)
+        return Segments(self.templated_file, *child_segments)
+
+    def first(self, *predicates) -> Optional[BaseSegment]:
+        """Returns the first segment (if any) that satisfies the predicates."""
+        cp = _CompositePredicate(*predicates)
+        for s in self:
+            if cp(s):
+                return s
+        return None
```

Comment:
feel like by default if no predicates are defined then this should just return the first segment in the tuple?

Reply:
That should happen already, because `_CompositePredicate.__call__()` returns `True` if there are no predicates. `return None` is for the case where no predicate is satisfied. I'll add a comment to clarify this.